### PR TITLE
Adding unittests for scribbles transforms in ./monailabel/scribbles/transforms.py

### LINKED ISSUE #324: Scribbles - Add unittests for scribbles transforms under ./monailabel/scribbles/transforms.py
As discussed in status meeting, this is to discuss and track progress of unittesting for scribbles transforms. 

@SachidanandAlle, @aihsani when I get to this task,  I may need to discuss for some guidance on how to approach this. At the moment, this is only focused  on scribbles transforms as these are core functions used for scribbles.

### PR DESCRIPTION
**Pull request by masadcv**
Fixes #324 
Adding unittests for the following scribbles transfroms from ./monailabel/scribbles/transforms.py
- `AddBackgroundScribblesFromROId`
- `ApplyGraphCutOptimisationd`
- `ApplyISegGraphCutPostProcd`
- `ApplySimpleCRFOptimisationd`
- ` MakeISegUnaryd`
- `MakeLikelihoodFromScribblesHistogramd`

Since these transforms use all methods currently implemented in ./monailabel/scribbles/utils.py therefore it indirectly covers those functions as well.

Inside the tests, I implemented a simple synthetic data generation function (based on monai's synthetic data functions) - that "simulates" the input data. This is used throughout the tests to verify the mathematical correctness of implemented methods. 

These tests cover different aspects of the above transforms, these include checking:
- 2D/3D input/output data variation
- Input arguments
- Output shapes
- Output comparison with expected output (checking mathematical correctness of algorithm applied)

Checklist:
- [x] All tests passing locally

Signed-off-by: masadcv <muhammad.asad@kcl.ac.uk>

### CODE REVIEW

File: tests/unit/scribbles/test_transforms.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        mask[
+            :,
+            test_input["roi"][0] : test_input["roi"][1],
+            test_input["roi"][2] : test_input["roi"][3],
+            test_input["roi"][4] : test_input["roi"][5],
+        ] = True
+        target = result["scribbles"]
+        target[~mask] = input_param["scribbles_bg_label"]
+
+        np.testing.assert_allclose(target, mask.astype(np.float32) * result["scribbles"], rtol=1e-4)
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:244]: isn't the target integer results? why is this comparison done in `float32` with a tolerance?

is the transform here not deterministic?

File: tests/unit/scribbles/test_transforms.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+import numpy as np
+from parameterized import parameterized
+
+from monailabel.scribbles.transforms import (
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:17]: These transform (along with the unit tests here) should be moved to MONAI core soon

File: tests/unit/scribbles/test_transforms.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    def test_add_bg_roi_transform(self, input_param, test_input, expected_shape):
+        result = AddBackgroundScribblesFromROId(**input_param)(test_input)
+        mask = result["scribbles"].astype(bool)
+        mask[
+            :,
+            test_input["roi"][0] : test_input["roi"][1],
+            test_input["roi"][2] : test_input["roi"][3],
+            test_input["roi"][4] : test_input["roi"][5],
+        ] = True
+        target = result["scribbles"]
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:241]: `target` should be something predetermined before the test is run and/or independent from the output of `AddBackgroundScribblesFromROId` so you can compare the `result["scribbles"]` against, not something derived from the result of `AddBackgroundScribblesFromROId`

Comment (aihsani) [reply tests/unit/scribbles/test_transforms.py:241]: e.g. `target` it should come in along with `expected_shape`

**Review by aihsani — CHANGES_REQUESTED** [review]

File: tests/unit/scribbles/test_transforms.py (hunk 3 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py:241]: You are right. Target needs to be built directly from roi input and then compared against output. I will correct this.

File: tests/unit/scribbles/test_transforms.py (hunk 1 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py:244]: This transform works on a "scribbles" input data, which is received from slicer as fp32 input. You are right that this could be done with integers, but it makes it easier to mix scribbles with logits and input image volume if it is fp32.

The transform, therefore, does not impose conversion of scribbles to integers or float32. It just adds values on top of what is already received (be it interger or fp32). 

Maybe there is a better way to be less ambiguous here? Perhaps two tests with 1) integers and 2) fp32 as input?

File: tests/unit/scribbles/test_transforms.py (hunk 3 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py]: Fixed this with using original input and making the output target on the fly.

File: tests/unit/scribbles/test_transforms.py (hunk 1 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py]: I have updated with two points:

1. Compare output directly without conversion to any data type
       `np.testing.assert_allclose(input_data, result["scribbles"], rtol=1e-4)`
2. Run the transform once with fp32 input and again with integer input. 

Let me know if this addresses your comment?

File: tests/unit/scribbles/test_transforms.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        np.testing.assert_equal(input_data, result["scribbles"])
+        self.assertTupleEqual(expected_shape, result["scribbles"].shape)
+        self.assertTupleEqual(test_input["scribbles"].shape, result["scribbles"].shape)
+
+    @parameterized.expand(TEST_CASE_OPTIM_TX)
+    def test_optimisation_transforms(self, input_param, test_input, output, expected_shape):
+        input_param.update({"post_proc_label": "pred"})
+        for current_tx in [ApplyGraphCutOptimisationd, ApplySimpleCRFOptimisationd]:
+            result = current_tx(**input_param)(test_input)
+            np.testing.assert_allclose(output["target"], result["pred"], rtol=1e-4)
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:271]: should this be considered also for a more deterministic `assert_equal`?

File: tests/unit/scribbles/test_transforms.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+            self.assertTupleEqual(expected_shape, result["pred"].shape)
+
+    @parameterized.expand(TEST_CASE_ISEG_OPTIM_TX)
+    def test_interactive_graphcut_optimisation_transform(self, input_param, test_input, output, expected_shape):
+        input_param.update({"post_proc_label": "pred"})
+        result = ApplyISegGraphCutPostProcd(**input_param)(test_input)
+        np.testing.assert_allclose(output["target"], result["pred"], rtol=1e-4)
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:278]: is there a better alternative to `assert_allclose`?

File: tests/unit/scribbles/test_transforms.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        result = MakeISegUnaryd(**input_param)(test_input)
+
+        # make expected unary output
+        expected_result = output["target"].copy()
+        eps = np.finfo(expected_result.dtype).eps
+        expected_result[expected_result == 0] = eps
+        expected_result[expected_result == 1] = 1 - eps
+
+        # compare
+        np.testing.assert_allclose(expected_result, result["pred"], rtol=1e-4)
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:293]: <same a s above>

File: tests/unit/scribbles/test_transforms.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    @parameterized.expand(TEST_CASE_MAKE_LIKE_HIST_TX)
+    def test_make_likelihood_histogram(self, input_param, test_input, output, expected_shape):
+        input_param.update({"post_proc_label": "pred"})
+        result = MakeLikelihoodFromScribblesHistogramd(**input_param)(test_input)
+
+        # make expected output
+        expected_result = np.argmax(output["target"].copy(), axis=0)
+
+        # compare
+        np.testing.assert_allclose(expected_result, np.argmax(result["pred"], axis=0), rtol=1e-4)
```
Comment (aihsani) [inline tests/unit/scribbles/test_transforms.py:305]: <same as above>

**Review by aihsani — APPROVED** [review]

File: tests/unit/scribbles/test_transforms.py (hunk 5 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py:278]: Since these are running optimisers inside, they cannot be guaranteed to have deterministic output. I will check with assert_equal or perhaps we can lower the tolerance.

Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py]: Okay, I have switched these to equal and all seems to work for this simple examples. In future if these pose an issue, we can switch to allclose with a tolerance.

File: tests/unit/scribbles/test_transforms.py (hunk 2 above)
Comment (masadcv) [reply tests/unit/scribbles/test_transforms.py:17]: Agreed!

**Review by aihsani — APPROVED** [review]

**Comment by masadcv** [comment]
Thanks!
